File: MonoStack.py
from collections import deque
def nle(A):
    stack = deque()
    nle = [-1] * len(A)

    for i in range(len(A)):
        while stack and A[stack[-1]] > A[i]:
            x = stack.pop()
            nle[x] = i
        stack.append(i)

    return nle


def ple(A):
    stack = deque()
    ple = [-1] * len(A)

    for i in range(len(A)):
        while stack and A[stack[-1]] > A[i]:
            stack.pop()
        ple[i] = stack[-1] if stack else -1
        stack.append(i)

    return ple


print(nle([1, 5, 6, 4, 0]) == [4, 3, 3, 4, -1])
print(ple([1, 5, 6, 4, 0]) == [-1, 0, 1, 0, -1])

File: Java/MonoStack.java
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.Stack;

public class MonoStack {

    public static int[] nle(int[] A) {
        Deque<Integer> stack = new ArrayDeque<>();
        int[] nle = new int[A.length];
        Arrays.fill(nle, -1);

        for (int i = 0; i < A.length; i++) {
            while (!stack.isEmpty() && A[stack.peek()] > A[i]) {
                int x = stack.pop();
                nle[x] = i;
            }
            stack.push(i);
        }

        return nle;
    }

    public static int[] ple(int[] A) {
        Stack<Integer> stack = new Stack<>();
        int[] ple = new int[A.length];
        Arrays.fill(ple, -1);

        for (int i = 0; i < A.length; i++) {
            while (!stack.empty() && A[stack.peek()] > A[i]) {
                stack.pop();
            }
            ple[i] = stack.empty() ? -1 : stack.peek();
            stack.push(i);
        }

        return ple;
    }

    public static void main(String[] args) {
        System.out.println(Arrays.equals(nle(new int[] {1, 5, 6, 4, 0}), new int[] {4, 3, 3, 4, -1}));
        System.out.println(Arrays.equals(ple(new int[] {1, 5, 6, 4, 0}), new int[] {-1, 0, 1, 0, -1}));
    }
}
